rcl/include/rcl/service.h

```diff
@@ -0,0 +1,318 @@
+ *
+ *    #include <rosidl_generator_c/service_type_support.h>
+ *    #include <example_interfaces/srv/add_two_ints.h>
+ *    rosidl_service_type_support_t * ts =
+ *      ROSIDL_GET_SERVICE_TYPE_SUPPORT(example_interfaces, AddTwoInts);
+ *
+ * For C++ a template function is used:
+ *
+ *    #include <rosidl_generator_cpp/service_type_support.hpp>
+ *    #include <example_interfaces/srv/add_two_ints.h>
+ *    rosidl_service_type_support_t * ts = rosidl_generator_cpp::get_service_type_support_handle<
+ *      example_interfaces::srv::AddTwoInts>();
+ *
+ * The rosidl_service_type_support_t object contains service type specific
+ * information used to send or take requests and responses.
+ *
+ * \TODO(wjwwood) update this once we've come up with an official scheme.
+ * The service name must be a non-empty string which follows the service/topic naming
+ * format.
+ *
+ * The options struct allows the user to set the quality of service settings as
+ * well as a custom allocator which is used when initializing/finalizing the
+ * client to allocate space for incidentals, e.g. the service name string.
+ *
+ * Expected usage (for C services):
+ *
+ *    #include <rcl/rcl.h>
+ *    #include <rosidl_generator_c/service_type_support.h>
+ *    #include <example_interfaces/srv/add_two_ints.h>
+ *
+ *    rcl_node_t node = rcl_get_zero_initialized_node();
+ *    rcl_node_options_t node_ops = rcl_node_get_default_options();
+ *    rcl_ret_t ret = rcl_node_init(&node, "node_name", &node_ops);
+ *    // ... error handling
+ *    rosidl_service_type_support_t * ts = ROSIDL_GET_SERVICE_TYPE_SUPPORT(
+ *      example_interfaces, AddTwoInts);
+ *    rcl_service_t service = rcl_get_zero_initialized_service();
+ *    rcl_service_options_t service_ops = rcl_service_get_default_options();
+ *    ret = rcl_service_init(&service, &node, ts, "add_two_ints", &service_ops);
+ *    // ... error handling, and on shutdown do finalization:
+ *    ret = rcl_service_fini(&service, &node);
+ *    // ... error handling for rcl_service_fini()
+ *    ret = rcl_node_fini(&node);
+ *    // ... error handling for rcl_node_fini()
+ *
+ * \param[out] service preallocated service structure
+ * \param[in] node valid rcl node handle
+ * \param[in] type_support type support object for the service's type
+ * \param[in] service_name the name of the service
+ * \param[in] options service options, including quality of service settings
+ * \return RCL_RET_OK if service was initialized successfully, or
+ *         RCL_RET_INVALID_ARGUMENT if any arugments are invalid, or
+ *         RCL_RET_BAD_ALLOC if allocating memory failed, or
+ *         RCL_RET_ERROR if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_service_init(
+  rcl_service_t * service,
+  const rcl_node_t * node,
+  const rosidl_service_type_support_t * type_support,
+  const char * service_name,
+  const rcl_service_options_t * options);
+
+/// Deinitialize a rcl_service_t.
```

Comment:
I agree. I was actually just wondering why we use `fini`, which isn't really a shortened version of `finalize`, but rather French/German for "finished" or "to terminate". Which is appropriate but not strictly English. I guess it's like saying "fin" at the end of a film. We've co-opted it :smile: